csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -7,13 +7,64 @@
 // clang-format on
 #include <alias_analysis.h>
 #include <debug.h>
+#include <ir/iostream.h> // for operator<<(ostream&, TensorView*)
 #include <ir/utils.h>
 #include <ops/alias.h>
 #include <options.h>
 #include <preseg_passes/mark_aliases_prepare.h>
 
 namespace nvfuser::preseg_passes {
 
+namespace {
+
+struct Use {
+  TensorView* use_of;
+  Expr* user;
+
+  bool operator<(const Use& other) const {
+    return std::make_pair(use_of, user) <
+        std::make_pair(other.use_of, other.user);
+  }
+
+  bool operator==(const Use& other) const {
+    return use_of == other.use_of && user == other.user;
+  }
+};
+
+// A helper function that walks up from `out` until reaching a non-meta op or a
+// fusion input. Returns where it stops.
+Use findUseToSegment(
+    TensorView* out,
+    const AliasAnalysisResult& analysis,
+    const std::unordered_set<Expr*>& used_by_non_aliases) {
+  Expr* user = nullptr;
+  while (true) {
+    Expr* def = out->definition();
+    if (analysis.getRoot(out) == nullptr || used_by_non_aliases.count(def)) {
+      return {out, user};
+    }
+    out = def->input(0)->as<TensorView>();
+    user = def;
+  }
+}
+
+// Collects all expressions that are (transitively) used by non-alias
+// TensorViews.
+std::unordered_set<Expr*> exprsUsedByNonAliases(
+    const AliasAnalysisResult& analysis,
+    Fusion* fusion) {
+  std::vector<Val*> non_aliases;
+  for (TensorView* tv : ir_utils::allTvs(fusion)) {
+    if (analysis.getRoot(tv) == nullptr) {
+      non_aliases.push_back(tv);
+    }
+  }
+  std::vector<Expr*> used_by_non_aliases = StmtSort::getExprsTo(non_aliases);
```

Comment:
IIUC, this `used_by_non_aliases` means `exprs_producing_non_alias`.

Reply:
I'll rename this to `depended_by_non_aliases`. I wanted to emphasize on transitivity, because it's not only the direct user that I care about.